### ISSUE TITLE
XHR error - M3U8 File for live stream is not being created 

### ISSUE THREAD

**Issue by guymass** [WWBN/AVideo#8501]
This is what I get when I try to view my live stream or any other:

![Screenshot 5784-01-18 at 14 53 54](https://github.com/WWBN/AVideo/assets/58882903/c97651d0-e682-4424-b840-5365bcfe4e31)

![Screenshot 5784-01-18 at 14 49 21](https://github.com/WWBN/AVideo/assets/58882903/91192382-d5eb-4a8d-9f7d-486964063da6)

Live stream is being encoded eventually just fine and is shown on site, but I cannot view my live stream anymore not sure why.

Everything is updated, cleared caches, nothing changed in my config files as far  as I can tell.
I see a message in log indicating the m3u8 file url has not been created and its waiting 5 seconds this being repeated.
I don't see anything being created under the /HLS/live directory, nginx has permissions to write to it. in general IDK yet what is the problem as I cannot view my live streams online.

**Comment by guymass** [comment]
verifying token https://alive528.com/plugin/Live/verifyToken.json.php?token=WWtoY0NvR2FZYU5QUThtcXhQb0FPOU1rK0hrdEdGUERvV2FhUU00dmZpYUZ>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: NGINX Live::on_publish end SCRIPT_NAME: /plugin/Live/on_publish.php
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: NGINX Live::on_publish YPTSocket start  (/usr/bin/php /var/www/html/AVideo/plugin/Live/on_publish_socket_notification.php '3' 'https:/>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: execAsync Linux: /usr/bin/php /var/www/html/AVideo/plugin/Live/on_publish_socket_notification.php '3' 'https://alive528.com:8443/live/>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: NGINX Live::on_publish YPTSocket end 2744484 SCRIPT_NAME: /plugin/Live/on_publish.php
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: Cache::isREQUEST_URIWhitelisted: (plugin/Live/verifyToken.json.php) is whitelisted SCRIPT_NAME: /plugin/Live/verifyToken.json.php
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: Live::verifyToken.json.php  {"salt":"5f9b7062dbe7b","timezone":"Asia\/Jerusalem","time":1696333160,"timeout":1696333220} SCRIPT_NAME: >
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: getRTMPLinkWithOutKey error ["_1696333160",false] SCRIPT_NAME: /plugin/Live/verifyToken.json.php
[03-Oct-2023 14:39:20 Asia/Jerusalem] Control.json.php verification respond content {"error":false,"msg":"","token":"WWtoY0NvR2FZYU5QUThtcXhQb0FPOU1rK0hrdEdGUERvV2FhUU00dmZpYUZ5Z0xYazcxQnUyQ>
[03-Oct-2023 14:39:20 Asia/Jerusalem] Control.json.php token is correct
[03-Oct-2023 14:39:20 Asia/Jerusalem] Control.json.php finish {"error":false,"msg":"","streamerURL":"https:\/\/alive528.com\/","token":"WWtoY0NvR2FZYU5QUThtcXhQb0FPOU1rK0hrdEdGUERvV2FhUU00dm>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: Live:controlRecording https://alive528.com/plugin/Live/standAloneFiles/control.json.php?command=record_start&app=live&name=6463c5bac68>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: Live:controlRecording start record is ready /var/www/tmp/6463c5bac6883-03-Oct-23-14:39:20.flv SCRIPT_NAME: /var/www/html/AVideo/plugin>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: NGINX Live::on_publish_socket_notification start SCRIPT_NAME: /var/www/html/AVideo/plugin/Live/on_publish_socket_notification.php
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: NGINX Live::on_publish_socket_notification (https://alive528.com:8443/live/6463c5bac6883.m3u8) SCRIPT_NAME: /var/www/html/AVideo/plugi>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: Live::getStatsObject[0] 5: forced to be recreated SCRIPT_NAME: /var/www/html/AVideo/plugin/Live/on_publish_socket_notification.php
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: isLiveAndIsReadyFromKey the m3u8 file is not present https://alive528.com:8443/live/6463c5bac6883.m3u8 [{"file":"\/var\/www\/html\/AVi>
[03-Oct-2023 14:39:20 Asia/Jerusalem] AVideoLog::DEBUG: live is not ready [11] request again in 5 seconds (https://alive528.com:8443/live/6463c5bac6883.m3u8) SCRIPT_NAME: /var/www/html/AVide>
[03-Oct-2023 14:39:23 Asia/Jerusalem] AVideoLog::DEBUG: isLiveAndIsReadyFromKey the m3u8 file is not present https://alive528.com:8443/live/6463c5bac6883.m3u8 [{"file":"\/var\/www\/html\/AVi>
[03-Oct-2023 14:39:23 Asia/Jerusalem] AVideoLog::DEBUG: Page generated in 0.6641 seconds. User: Not Logged - No Cache (/page/10?page=4&lang=sv_SE) FROM: 52.167.144.137 Browser: Mozilla/5.0 A>
[03-Oct-2023 14:39:24 Asia/Jerusalem] AVideoLog::DEBUG: Time: 0.310s   | Limit: 0.2s | Location: /var/www/html/AVideo/view/modeYoutube.php Line 120 [navbar.php] USER_AGENT=Mozilla/5.0 AppleW>
[03-Oct-2023 14:39:24 Asia/Jerusalem] AVideoLog::DEBUG: Page generated in 0.4892 seconds. User: Not Logged - No Cache (/v/15395) FROM: 52.167.144.137 Browser: Mozilla/5.0 AppleWebKit/537.36 >
[03-Oct-2023 14:39:25 Asia/Jerusalem] AVideoLog::DEBUG: CustomizeUser::canShareVideosFromVideo canShareVideosFromUser(1559) = false SCRIPT_NAME: /view/index.php

**Comment by guymass** [comment]
And I see this error although my URL should point to my network encoders server that manages the encoders. This was always set like that so im not sure why now it gives an error.

(https://encoder.alive528.com/) is not configured properly This slow down your site a lot, disable the option useEncoderNetworkRecomendation in your CustomizeAdvanced plugin

**Comment by guymass** [comment]
![Screenshot 5784-01-18 at 15 28 03](https://github.com/WWBN/AVideo/assets/58882903/d159f410-ee6b-4146-b6f0-65851b8bb21b)
![Screenshot 5784-01-18 at 15 27 16](https://github.com/WWBN/AVideo/assets/58882903/d70cd4a6-daf6-4a62-b5f3-71a93de59a51)

**Comment by guymass** [comment]
The issue is the live stream URL - https://alive528.com/live/0/FE-Streamer
doesn't work, I get errors trying to open or create a missing m3u8 file.

![Screenshot 5784-01-18 at 19 20 23](https://github.com/WWBN/AVideo/assets/58882903/6c65abee-d641-4edc-9cf6-85bd187b4f30)

[03-Oct-2023 19:20:35 Asia/Jerusalem] AVideoLog::DEBUG: isLiveAndIsReadyFromKey the m3u8 file is not present https://alive528.com:8443/live/6463c5bac6883.m3u8 [{"file":"\/var\/www\/html\/AVideo\/plugin\/Live\/Live.php","line":2510,"function":"isLiveAndIsReadyFromKey","class":"Live","type":"::"},{"file":"\/var\/www\/html\/AVideo\/plugin\/Live\/getImage.php","line":78,"function":"isLive","class":"Live","type":"::"}] SCRIPT_NAME: /plugin/Live/getImage.php

**Comment by GreenDragonTavern** [comment]
@guymass - Fixed or adaptive bit rate on live stream?

**Comment by guymass** [comment]
I have set adaptive here is my nginx configuration:

rtmp {
    access_log /var/log/nginx/rtmp_access.log;

            server {
                    listen 178.162.220.42:1935;
                    chunk_size 8000;
                    #creates our "live" full-resolution HLS videostream from our incomi>
                    application live {
                         live on;
                        allow publish all;
                        allow play all;
                        interleave off;
                        meta on;
                        wait_key on;
                        wait_video on;
                        idle_streams off;
                        sync 300ms;
                        session_relay on;
                        max_connections 1000;


                ## == DASH == ##
                dash off;
                # dash_nested on;
                # dash_path /var/livestream/dash;
                # dash_fragment 10s;
                # dash_playlist_length 60s;
                # dash_cleanup on;

		
    on_publish http://178.162.220.42/plugin/Live/on_publish.php;
    on_publish_done http://178.162.220.42/plugin/Live/on_publish_done.php;
    on_play http://178.162.220.42/plugin/Live/on_play.php;
    on_record_done http://178.162.220.42/plugin/Live/on_record_done.php;


        #ADAPTATIVE MODE


exec ffmpeg -i rtmp://178.162.220.42/live/$name 
                                -c:a aac -strict -2 -b:a 96k -c:v libx264 -vf scale=-2:240 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 400k -maxrate 700k -bufsize 1200k -b:a 96k -r 20 -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp://178.162.220.42/adaptive/$name_low
                                -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:480 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 1200k -maxrate 2100k -bufsize 2400k -b:a 128k -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp://178.162.220.42/adaptive/$name_mid  
                                -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:720 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 2400k -maxrate 4200k -bufsize 100000k -b:a 128k -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp://178.162.220.42/adaptive/$name_hi; 

                           recorder video{
                                record all manual;
                                record_path /var/www/tmp;
                                record_notify on;
                                record_max_size 4196M;
				#record_max_frames 30;
                                record_suffix -%d-%b-%y-%T.flv;
                            }
                    }


                #ADAPTATIVE MODE
                application adaptive {
                    live on;
                    hls on;
#                    hls_fragment_naming system;
                    hls_path /HLS/live;
                    hls_nested on;
                    hls_playlist_length 120m;
                    allow play all;
                    allow publish 127.0.0.1;
                    allow publish 84.16.238.224;
                    allow publish 178.162.220.42;
                    deny publish all;

                    hls_variant _low BANDWIDTH=400000;
                    hls_variant _mid BANDWIDTH=2100000;
                    hls_variant _hi  BANDWIDTH=4200000;

#                    hls_base_url https://alive528.com:8443/live/;
                }
            }
    }

**Comment by guymass** [comment]
I also reconfigured and compiled again my FFMPEG with these parameters:
./configure --host-ldflags=-L/usr/local/lib --enable-gpl --enable-libgsm --enable-libxvid \             
--enable-libmp3lame --enable-libvorbis --enable-shared

**Comment by GreenDragonTavern** [comment]
You will never get direct viewing of live ie; https://AVideoDomain.com:8443/live/UserLiveKey.m3u8 with adaptive bit rates without mods. Reason is simple, your nginx config is creating separate directories for each bit rate. This is a bug (and what is adaptive flag purpose in Live Plugin?) thus making adaptive bitrate worthless. If your userbase is utilizing these direct links, probably better off using fixed bitrate as AVideo will work without modification or modify AVideo Live plugin to reflect correct bitrate path you'd like to stream on Player URL parameter.

**Comment by DanielnetoDotCom** [comment]
For live streams, refer to the following guide: https://github.com/WWBN/AVideo/wiki/What-to-do-if-your-live-does-not-play

**Comment by GreenDragonTavern** [comment]
@DanielnetoDotCom - If using fixed bitrate works like a champ, adaptive not true and does not work as advertised.

**Comment by DanielnetoDotCom** [comment]
@GreenDragonTavern what do you mean by fix bitrates?

**Comment by GreenDragonTavern** [comment]
> @GreenDragonTavern what do you mean by fix bitrates?
> 
> I do have adaptive in many customers, and it is working great.

Sure if using AVideo link ie; https://AVideo.com/live/0/UserChannel/1 as example. What about Player URL only and adaptive bitrates?

**Comment by DanielnetoDotCom** [comment]
I am sorry I am still not sure what do you mean

**Comment by GreenDragonTavern** [comment]
When a live starts and AVideo displays notification or shows up on front page, all of those links are contained within AVideo. In other words AVideo code has total control of displaying said live stream. Now the Player URL link is just that, a link to the nginx server for any html5 player to play. It's the Player URL that fails using adaptive mode.

**Comment by guymass** [comment]
I think he is right because I used to have different urls once and adaptive is what I have in the last few years, however I did lately limited my encoders to only encode 720p because I wanted to test just using that resolution for a while. Maybe this is also what causing my issue?
![Screenshot 5784-01-18 at 20 56 42](https://github.com/WWBN/AVideo/assets/58882903/4689bedb-3055-4510-982a-5f0a03bd534d)

**Comment by guymass** [comment]
Ok I removed the checkbox from the adaptive in the Live plugin and set my nginx  like the following and now I can view my live stream. So adaptive is currently not working as far as I can tell.

application live {
                        live on; 
                        allow publish all;

    on_publish http://178.162.220.42/plugin/Live/on_publish.php;
    on_publish_done http://178.162.220.42/plugin/Live/on_publish_done.php;
    on_play http://178.162.220.42/plugin/Live/on_play.php;
    on_record_done http://178.162.220.42/plugin/Live/on_record_done.php;
        #ADAPTATIVE MODE
                            # exec ffmpeg -i rtmp://178.162.220.42/live/$name 
                            #     -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:240 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 400k -maxrate 700k -bufsize 1200k -b:a 96k -r 20 -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp://178.162.220.42/adaptive/$name_low  
                            #     -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:480 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 1200k -maxrate 2100k -bufsize 2400k -b:a 128k -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp://178.162.220.42/adaptive/$name_mid  
                            #     -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:720 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 2400k -maxrate 4200k -bufsize 100000k -b:a 128k -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp://178.162.220.42/adaptive/$name_hi;

    exec ffmpeg -i rtmp://178.162.220.42/live/$name 
    -c:a aac -strict -2 -b:a 96k -c:v libx264 -vf scale=-2:240 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 400k -maxrate 700k -bufsize 1200k -b:a 96k -r 20 -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp:/178.162.220.42/adaptive/$name_low
    -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:480 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 1200k -maxrate 2100k -bufsize 2400k -b:a 128k -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp:/178.162.220.42/adaptive/$name_mid  
    -c:a aac -strict -2 -b:a 128k -c:v libx264 -vf scale=-2:720 -g 48 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 2400k -maxrate 4200k -bufsize 100000k -b:a 128k -f hls -hls_time 6 -hls_list_size 0 -f flv rtmp:/178.162.220.42/adaptive/$name_hi; 

    recorder video{
       record all manual;
       record_path /var/www/tmp;
       record_notify on;
       record_max_size 4096M;
       #record_max_frames 30;
       record_suffix -%d-%b-%y-%T.flv;
       }
    }

**Comment by guymass** [comment]
Now its running in loop only showing the first few min and then the player hangs, but my live stream still goes on. When I refresh it just starts from the last part, but doesn't show me the current state of the live video stream. 

So I checked my /HLS/live directory and I have seen a directory created with my key, so I removed that and now im back to 0 again cannot watch my live stream. So in short my NGINX doesn't create the folders it needs when I changed my configuration.

So which part of the NGINX is responsible for creating thtose folders so theat the index.m3u8 file can be created in them?

**Comment by GreenDragonTavern** [comment]
Your nginx is a kluge. If you're going to go fixed, why are you creating 3 bitrates for an adaptive application that is not necessary? Rem out code for exec ffmpeg and entire adaptive application.

**Comment by guymass** [comment]
How is this:

rtmp {
    access_log /var/log/nginx/rtmp_access.log; 
            server {
                    listen 178.162.220.42:1935;
                    allow play all;

               application live {

                    live on;
		    allow play all;
		    allow publish all;
                    hls on;
		    hls_type live; # guy added
		    #hls_fragment_naming system;
		    hls_nested on;
		    hls_path /HLS/live;
		    hls_playlist_length 120m;
		    hls_fragment 4s;				
		    hls_sync 100ms;
		    hls_cleanup on;
			
			interleave off;
			meta on;
			wait_key on;
			wait_video on;
			idle_streams off;
			sync 300ms;
			session_relay on;
			allow publish all;
			max_connections 10000;


		## == DASH == ##
		dash off;
		# dash_nested on;
		# dash_path /var/livestream/dash;
		# dash_fragment 10s;
		# dash_playlist_length 60s;
		# dash_cleanup on;


    on_publish http://178.162.220.42/plugin/Live/on_publish.php;
    on_publish_done http://178.162.220.42/plugin/Live/on_publish_done.php;
    on_play http://178.162.220.42/plugin/Live/on_play.php;
    on_record_done http://178.162.220.42/plugin/Live/on_record_done.php;
                            recorder video{
                                record all;
                                record_path /var/www/tmp;
                                record_notify on;
                                record_max_size 4192M;
				#record_max_frames 30; 
                                record_suffix -%d-%b-%y-%T.flv;
                            }
                    }
            }
    }

**Comment by GreenDragonTavern** [comment]
Here's a copy of template from ../plugin/Live/install/nginx.conf which is configured for single fixed stream but also has all necessary to go adaptive after bug is fixed. 
```
    rtmp {
            server {
                    listen 1935;
                    buflen 1000ms;
                    chunk_size 4096;
                    ping_timeout 30s;
                    ### Use case option (max_streams; default is 32 )
                    #max_streams 10;
                    allow play all;
                    #creates our "live" full-resolution HLS videostream from our incoming encoder stream and tells where to put the HLS video manifest and video fragments
                    application live {
                            allow play all;
                            live on;
                            hls on;
                            hls_nested on;
                            hls_path /HLS/live;
                            hls_playlist_length 60m;
                            hls_fragment 4s;
                            drop_idle_publisher 30s;
                            sync 500ms;
                            #Experimental. Force dropped stream, or ended stream from being watched. (idle_streams)
                            #idle_streams off;
                            on_publish http://localhost/AVideo/plugin/Live/on_publish.php;
                            on_publish_done http://localhost/AVideo/plugin/Live/on_publish_done.php;
                            on_play http://localhost/AVideo/plugin/Live/on_play.php;
                            on_record_done http://localhost/AVideo/plugin/Live/on_record_done.php;
                            
                            #exec ffmpeg -re -i rtmp://localhost/live/$name -c:v libx264 -preset veryfast -c:a copy -f hls -hls_time 5 -hls_list_size 0 -f flv rtmp://localhost/adaptive/$name_hi;
                            #exec ffmpeg -re -i rtmp://localhost/live/$name 
                            #    -c:v libx264 -vf scale=-2:240 -r 20 -g 40 -keyint_min 40 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 400k -maxrate 700k -bufsize 1400k -c:a aac -strict -2 -b:a 96k -f flv rtmp://localhost/adaptive/$name_low  
                            #    -c:v libx264 -vf scale=-2:480 -r 30 -g 60 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 1200k -maxrate 2100k -bufsize 4200k -c:a aac -strict -2 -b:a 128k -f flv rtmp://localhost/adaptive/$name_mid  
                            #    -c:v libx264 -vf scale=-2:720 -r 30 -g 60 -keyint_min 48 -sc_threshold 0 -bf 3 -b_strategy 2 -b:v 2400k -maxrate 3000k -bufsize 6000k -c:a aac -strict -2 -b:a 128k -f flv rtmp://localhost/adaptive/$name_hi;


                            #recorder video {
                            #    record all manual;
                            #    record_path /var/www/tmp;
                            #    record_notify on;
                            #    record_max_size 2048M; 
                            #    record_suffix -%d-%b-%y-%T.flv;
                            ###If live-stream is over 30FPS and you want to limit it to 30FPS recording to light the encoding task #record_max_frames 30;
                            #    ## record_max_frames 30;
                            #}
                    
                            ### Record Audio Separately ( For podcast )
                            #recorder audio {
                            #    record audio;
                            #    record_path /var/www/tmp;
                            #    record_max_size 1024M;
                            #    record_suffix -%d-%b-%y-%T.mp3;
                            #}
                    }

                    #application adaptive {
                    #    live on; 
                    #    hls on; 
                    #    hls_fragment 2s;
                    #    hls_path /HLS/live;
                    #    hls_nested on;
                    #    hls_playlist_length 10m;
                    #    allow play all;
                    #    allow publish 127.0.0.1;
                    #    deny publish all;

                    #    hls_variant _hi BANDWIDTH=264000,RESOLUTION=1280x720; # this is for line 34 only, do not uncomment it
                    #    hls_variant _low BANDWIDTH=900000;
                    #    hls_variant _mid BANDWIDTH=2400000;
                    #    hls_variant _hi  BANDWIDTH=3500000;
                    #}
            }
    }

```

**Comment by guymass** [comment]
OK I think I got it working now and also the encoding is working so I don't want to touch it again ahaha. Thank you @GreenDragonTavern  for your help. I think you are right currently adaptive is causing issues, however for the encoding to work I needed to leave the FFMPEG in there, otherwise encoding will not work for me. I will keep this post open till tomorrow just to make sure it all works well for all other users.
